File: dei-rs/crates/dei-e2e/fixtures/csharp/Healthy.cs
// Healthy, well-structured C# code

using System;
using System.Collections.Generic;

namespace TestApp.Services
{
    /// <summary>
    /// A simple, focused product repository
    /// </summary>
    public class ProductRepository
    {
        private readonly Dictionary<int, Product> _products;

        public ProductRepository()
        {
            _products = new Dictionary<int, Product>();
        }

        public void Add(Product product)
        {
            if (product == null)
                throw new ArgumentNullException(nameof(product));

            if (_products.ContainsKey(product.Id))
                throw new InvalidOperationException("Product already exists");

            _products[product.Id] = product;
        }

        public Product Get(int id)
        {
            return _products.TryGetValue(id, out var product) ? product : null;
        }

        public void Remove(int id)
        {
            _products.Remove(id);
        }

        public int Count()
        {
            return _products.Count;
        }

        public IEnumerable<Product> GetAll()
        {
            return _products.Values;
        }
    }

    public class Product
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public decimal Price { get; set; }
        public string Category { get; set; }

        public Product(int id, string name, decimal price, string category)
        {
            Id = id;
            Name = name;
            Price = price;
            Category = category;
        }

        public bool IsValid()
        {
            return !string.IsNullOrEmpty(Name) && Price > 0;
        }
    }
}

File: dei-rs/crates/dei-e2e/fixtures/rust/healthy.rs
//! Healthy, well-structured Rust code

use std::collections::HashMap;

/// A simple, focused user repository
pub struct UserRepository {
    users: HashMap<u64, User>,
}

impl UserRepository {
    pub fn new() -> Self {
        Self {
            users: HashMap::new(),
        }
    }

    pub fn add(&mut self, user: User) -> Result<(), String> {
        if self.users.contains_key(&user.id) {
            return Err("User already exists".to_string());
        }
        self.users.insert(user.id, user);
        Ok(())
    }

    pub fn get(&self, id: u64) -> Option<&User> {
        self.users.get(&id)
    }

    pub fn remove(&mut self, id: u64) -> Option<User> {
        self.users.remove(&id)
    }

    pub fn count(&self) -> usize {
        self.users.len()
    }
}

#[derive(Clone, Debug)]
pub struct User {
    pub id: u64,
    pub name: String,
    pub email: String,
}

impl User {
    pub fn new(id: u64, name: String, email: String) -> Self {
        Self { id, name, email }
    }

    pub fn validate_email(&self) -> bool {
        self.email.contains('@')
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_user_repository() {
        let mut repo = UserRepository::new();
        let user = User::new(1, "Alice".to_string(), "alice@example.com".to_string());
        assert!(repo.add(user).is_ok());
        assert_eq!(repo.count(), 1);
    }
}

